### ISSUE TITLE
On ObsCore profiles, Obscore extensions and registry matters

### ISSUE THREAD

**Issue by Bonnarel** [ivoa-std/ObsCoreExtensionForRadioData#73]
If we want to complete the basic ObsCore table by a set of new standard attributes (= columns) in the TAP_SCHEMA we can define this specific set of columns as an "ObsCore profile"

The profile is relevant to a "schema" in the tableset which should contain all the columns in whatever tables inside the schema. 

It is admitted that in the radio case the basic ObsCore parameters and the radio extension specific parameters will be hosted in two different tables belonging to the same "ivoa" schema. The recommendation  is that the basic ObsCore table is called ivoa.obscore and that the extension table is called ivoa.obscore_radio. 

The main reason is that the same TAP service may contain data in the radio domain and data outside this domain for which the extension  parameters are irrelevant.
So storing everything in the same table would imply that many rows in the table will show NULL values for the extension parameters.

---

But how do we help users to discover services which serve these two tables ?

Each standard data model has a standardID. This is the case of ObsCore, of EPN-TAP, of RegTAP....

From the registry point of view, the occurrence of an  ObsCore table itself in a service is recognized via the datamodel element of a service capability. This practice has the drawback  to match a datamodel with a service and not with the tables it serves. 

That's the reason why the practice changed and why EPN-TAP and RegTAP used another method.

This is summarized in  this recent IVOA note published by Markus : 

 https://ivoa.net/documents/Notes/TableReg/20240821/
 
 So if the model is serialized in a single table, let's set the standardID of this model as the value of the utype attribute of the table in the registry record 
 
 So if the model is serialized in several tables, let's set the standardID of this model as the value of the utype attribute of the schema grouping all these tables in the registry record.
 
 What can happen for the  ObsCore extension for radio data ? Obscore and it's extension are actually part of the same data model. So they have the same basic standardID "ivo://ivoa.net/std/ObsCore" and the  presence of the extension columns may be rendered by a fragment in the URI "ivo://ivoa.net/std/ObsCore#RadioEXt-1.0"
 
 
 Now where can we put this standardID in the registry ? 
Currently the thing is organized in two tables which are strongly related. So they are in the same "schema". However setting
 utype="ivo://ivoa.net/std/ObsCore#RadioEXt-1.0" on the schema is not appropriate. This would mean all the tables in the schema are dealing with radio data. But we may have dataset in the ObsCore table which are outside the radio domain. 
 
 Another solution could be to set the utype="ivo://ivoa.net/std/ObsCore#RadioEXt-1.0" on the ivoa.obscore_radio table only. 
 
 But this may be confusing and encourage users to try to query this table only which would be a nonsense. The obscore_radio table alone is useless Only a query on a natural join between obscore and obscore_radio makes sense.
 
 Hence the proposal to set the standardID utype on a "view" table defined as the natural join of the ObsCore and obscore_radio tables. If such a utype is discovered in a service we know that the schemac ontaining this view will also contain ObsCore and obscore_radio tables.  Of course in practice it may be more efficient to query the two tables by a direct join instead of this view. But the view is there to inform registry users that this service serves ObsCore with its radio extension.
 
 By the way the concept of specific profiles of ObsCore defined this way as a full set of columns is agnostic about in which real table we find the columns. Hence if we decide later to move some of the extension columns in the basic Obscore it's very easy to redefine the profile in a new version. This will not break anything.

**Comment by msdemlei** [comment]
Just to prevent possible confusion: The current draft says we should have "ivo://ivoa.net/std/ObsCore#radioExt-1.0" in the table's utype (incidentally: let's lowercase that identifier in the spec; I'd like to move away from relying on case-insensitive ivoids).

If I understand François' concern right, it is basically that he argues there is no point in separately discovering that table, because there are probably no conceivable science queries against that table alone.

The premise of that statement is probably right. I don't think I agree with the conclusion, though.  You see, I think only a prankster would have obs_radio without ivoa.obscore, and hence if you see a #radioExt-1.0-utyped table, you *know* you can run queries joining obs_radio and obscore on it.  If you don't want to rely on people not being praksters, we could add a "if you publish obs_radio, you MUST have ivoa.obscore, too" in the spec.

Against that, creating a view in order to have something usable without a join to me seems to set a few traps for a very limited benefit; for one, operators need to make sure to create these views every time they re-create one of the underlying tables, and if we follow this pattern for other extensions, this can become a chore even for a computer.

Also, it feels a bit odd when you want, say, a service with both a radio and a time extension: You would discover the two views but then use neither of them (assuming we don't want to in addition create a radio-and-time view, too, which to me is too horrible to even contemplate).

So... my vote is for keeping things as they are.

**Comment by Bonnarel** [comment]
Sorry for the delay, Markus.

> Just to prevent possible confusion: The current draft says we should have "ivo://ivoa.net/std/ObsCore#radioExt-1.0" in the table's utype (incidentally: let's lowercase that identifier in the spec; I'd like to move away from relying on case-insensitive ivoids).
> 
OK. Done.
> If I understand François' concern right, it is basically that he argues there is no point in separately discovering that table, because there are probably no conceivable science queries against that table alone.
> 
Right, Markus. But not only this. See below.

> The premise of that statement is probably right. I don't think I agree with the conclusion, though. You see, I think only a prankster would have obs_radio without ivoa.obscore, and hence if you see a #radioExt-1.0-utyped table, you _know_ you can run queries joining obs_radio and obscore on it. If you don't want to rely on people not being praksters, we could add a "if you publish obs_radio, you MUST have ivoa.obscore, too" in the spec.
> 
> Against that, creating a view in order to have something usable without a join to me seems to set a few traps for a very limited benefit; for one, operators need to make sure to create these views every time they re-create one of the underlying tables, and if we follow this pattern for other extensions, this can become a chore even for a computer.
>
  The important point is to have a way to identify in one spot that the service, (the ivoa schema in the service more accurately) actually contains radio data described by the root obscore + extension. If creating a view containing all attributes is a concern we can have an empty table instead. Only the list of attributes as a  consistent set.   And forget about the actual view building I was proposing.

Then the standard ID utype I was proposing would be set on this empty table. This would imply that both the root table and the radio table with extension attributes will be there in the service ivoa schema.

> Also, it feels a bit odd when you want, say, a service with both a radio and a time extension: You would discover the two views but then use neither of them (assuming we don't want to in addition create a radio-and-time view, too, which to me is too horrible to even contemplate).
> 
 But in that case I don't see the difference : we need an id for that part of the tables that uses both radio and time extension attributes (pulsar radio data, FRB, radio source time series, etc...)

Because finding out a service which contains both radio extension and time extension would not imply that the service delivers a specific pulsar subset which will contain both.

I think the final correct view will be a an extended set of classes and attributes required for datasets descriptions in all electromagnetic domains.

Among all these attributes we can define profiles. The current overall domain ObsCore set will be the root profile.

Then we can have the radio profile, the time profile, the time+radio (pulsar ?) profile, the heig profile, ... all containing the root attributes + the dedicated others

Each profile will have it's stantardID (root Obscore one + a dedicate fragment)

I propose to write some dummy scenarii to show how it could work.
>

> So... my vote is for keeping things as they are.

Current Proposed recommendation contains a version of the registry section with utype on  the  ivoa schema which I think now is wrong because it implies that only radio data are included in the service. So I think you mean go back to your previous proposal which was

 %\section{Registry Aspects}
 %\label{sec:registry}
 %
 %Services compliant with this specification are registered using
 %VODataService \citep{2021ivoa.spec.1102D} tablesets.  Compliant tables
 %use the utype
 %$$
 %\hbox{\nolinkurl{ivo://ivoa.net/std/obsradio#table-1.0}.}
 %$$
 %
 %While it is admitted that the table only sits in the tableset of the
 %embedding TAP service, implementors are urged to use a seperate registry
 %record with the main TAP service as an auxiliary capability
 %\citep{2019ivoa.spec.0520D}.  In this way, meaningful information
 %on coverage in space, time, and spectraum as per VODataService 1.2 can
 %be communicated to the Registry, which, again, data providers are urged
 %to do.  There is no expectation that the coverage information only
 %pertains to data with entries in \verb|ivoa.obs_radio|, i.e., it may be
 %a copy of the coverage of the full Obscore table.\todo{Is that
 %acceptable?  Or should we require pure radio coverage?}
 %
 %A sample registry record for an obs\_radio table comes with this
 %specification\footnote{\auxiliaryurl{sample-record.xml}}.
 %
 %To obtain access URLs of all TAP services that have compliant tables
 %together with their table names (which in this major version are fixed
 %to \verb|ivoa.obs_radio|), use a RegTAP \citep{2019ivoa.spec.1011D}
 %query like:
 %
 %\begin{lstlisting}
 %SELECT DISTINCT(access_url), table_name
 %FROM rr.res_table
 %  NATURAL JOIN rr.capability
 %  NATURAL JOIN rr.interface
 %WHERE
 %  standard_id LIKE 'ivo://ivoa.net/std/tap%'
 %  AND intf_role='std'
 %  AND table_utype LIKE 'ivo://ivoa.net/std/obsradio#table-1.%'
 %\end{lstlisting}

**Comment by msdemlei** [comment]
On Sun, Jan 12, 2025 at 02:52:23PM -0800, François Bonnarel wrote:
> > Just to prevent possible confusion: The current draft says we
> > should have "ivo://ivoa.net/std/ObsCore#radioExt-1.0" in the
> > table's utype (incidentally: let's lowercase that identifier in
> [...]
> The important point is to have a way to identify in one spot that
> the service, (the ivoa schema in the service more accurately)
> actually contains radio data described by the root obscore +
> extension. If creating a view containing all attributes is a

I would submit that "there is a table with the #radioext-1.0 utype in
this TAP service" is exactly that spot.  What do you see missing from
it?  Or is it some duplication that you are concerned about?

> > Also, it feels a bit odd when you want, say, a service with both
> > a radio and a time extension: You would discover the two views
> > but then use neither of them (assuming we don't want to in
> > addition create a radio-and-time view, too, which to me is too
> > horrible to even contemplate).
> >
> But in that case I don't see the difference : we need an id for
> that part of the tables that uses both radio and time extension
> attributes (pulsar radio data, FRB, radio source time series,
> etc...)

I'm afraid I quite get what you are saying here, and I think that's
mainly because I don't really understand what "that part of the
tables that uses both radio and time" would be.  Are you thinking of
a rowset here, i.e., something that would be returned if you say

  ivoa.obscore
  NATURAL JOIN ivoa.obs_radio
  NATURAL JOIN ivoa.obs_time

?

I would give you it would be (mildly) desirable to be able to
discover "This service has observations that have extra metadata in
*both* the radio and the time extensions, but I think we'd have to pay
a high price for building something like that, and I'd submit we
should first get single-column statistics addressed properly
[shameless plug: https://ivoa.net/documents/Notes/colstatnote/ is
still waiting for feedback and implementations] before even daring to
think about this kind of correlated statistics.

Or are you thinking about column sets?  If so, doesn't #radioext-1.0
uniquely specify a well-defined column set that's exactly what people
would want to discover ("Give me all services I can run my query
using the radio extension against")?

> Because finding out a service which contains both radio extension
> and time extension would not imply that the service delivers a
> specific pulsar subset which will contain both.

This would suggest you are thinking of row subsets, and as just said
I strongly suspect that'll be too hard for data discovery for years
to come.  I also don't think that's a very strong discovery case to
begin with: it's probably almost always faster to just query a few
extra services than to come up with a clever discovery condition.

> Among all these attributes we can define profiles. The current
> overall domain ObsCore set will be the root profile.
>
> Then we can have the radio profile, the time profile, the
> time+radio (pulsar ?) profile, the heig profile, ... all containing
> the root attributes + the dedicated others
>
> Each profile will have it's stantardID (root Obscore one + a
> dedicate fragment)

But that would be the combinatorial catastrophe I was always afraid
of.  I give you we probably won't have 10 extensions for quite a
while, but if we had them, we'd have 2^10=1024 such "profiles".  I,
for one, would consider that a nightmare.

> Current Proposed recommendation contains a version of the registry
> section with utype on  the  ivoa schema which I think now is wrong

"On the ivoa schema" would be wrong, right. But at least in my docs
it's the obs_radio table that the utype is on.  And I'd argue that's
exactly where it should be.

> because it implies that only radio data are included in the
> service. So I think you mean go back to your previous proposal
> which was

Uh... My checkout (git commit d50ff42d) still says

  Compliant *tables* use the utype
  ivo://ivoa.net/std/ObsCore#radioExt-1.0.

(emphasis mine), and the sample-record.xml also has the utype at the
right position.  If the document says the utype is on the schema
somewhere, that's a bug and obviously needs to be fixed.

Where do you see that?  [the doc repo is down for me right now, so I
looked at github HEAD, which I think ought to be at least as good...]

**Comment by Bonnarel** [comment]
> On Sun, Jan 12, 2025 at 02:52:23PM -0800, François Bonnarel wrote:
>  > Just to prevent possible confusion: The current draft says we
>  > should have "ivo://ivoa.net/std/ObsCore#radioExt-1.0" in the
>  > table's utype (incidentally: let's lowercase that identifier in
>  [...]
>  The important point is to have a way to identify in one spot that
>  the service, (the ivoa schema in the service more accurately)
>  actually contains radio data described by the root obscore +
>  extension. If creating a view containing all attributes is a
> I would submit that "there is a table with the #radioext-1.0 utype in
> this TAP service" is exactly that spot.  What do you see missing from
> it?  Or is it some duplication that you are concerned about?

If you find out this table you still have to check if there is a root ObsCore table in the same schema

>  > Also, it feels a bit odd when you want, say, a service with both
>  > a radio and a time extension: You would discover the two views
>  > but then use neither of them (assuming we don't want to in
>  > addition create a radio-and-time view, too, which to me is too
>  > horrible to even contemplate).
>  >
>  But in that case I don't see the difference : we need an id for
>  that part of the tables that uses both radio and time extension
>  attributes (pulsar radio data, FRB, radio source time series,
>  etc...)
> I'm afraid I quite get what you are saying here, and I think that's
> mainly because I don't really understand what "that part of the
> tables that uses both radio and time" would be.  Are you thinking of
> a rowset here, i.e., something that would be returned if you say
> 
>   ivoa.obscore
>   NATURAL JOIN ivoa.obs_radio
>   NATURAL JOIN ivoa.obs_time
> 
> ?

Yes something like that, if you want. For pulsar radio data see this preliminary work :  https://github.com/ivoa/PulsarRadioDataAccess. A table describing these data will be really specific.

> 
> I would give you it would be (mildly) desirable to be able to
> discover "This service has observations that have extra metadata in
> *both* the radio and the time extensions, but I think we'd have to pay
> a high price for building something like that, and I'd submit we
> should first get single-column statistics addressed properly
> [shameless plug: https://ivoa.net/documents/Notes/colstatnote/ is
> still waiting for feedback and implementations] before even daring to
> think about this kind of correlated statistics.
> 
> Or are you thinking about column sets?  If so, doesn't #radioext-1.0
> uniquely specify a well-defined column set that's exactly what people
> would want to discover ("Give me all services I can run my query
> using the radio extension against")?
>  Because finding out a service which contains both radio extension
>  and time extension would not imply that the service delivers a
>  specific pulsar subset which will contain both.
> This would suggest you are thinking of row subsets, and as just said
> I strongly suspect that'll be too hard for data discovery for years
> to come.  I also don't think that's a very strong discovery case to
> begin with: it's probably almost always faster to just query a few
> extra services than to come up with a clever discovery condition.
>  Among all these attributes we can define profiles. The current
>  overall domain ObsCore set will be the root profile.
> 
>  Then we can have the radio profile, the time profile, the
>  time+radio (pulsar ?) profile, the heig profile, ... all containing
>  the root attributes + the dedicated others
> 
>  Each profile will have it's stantardID (root Obscore one + a
>  dedicate fragment)
> But that would be the combinatorial catastrophe I was always afraid
> of.  I give you we probably won't have 10 extensions for quite a
> while, but if we had them, we'd have 2^10=1024 such "profiles".  I,
> for one, would consider that a nightmare.

No. If we speak about spectral domains we will never combine them (Heig and radio will never be combined)
The only point is time extension + some specific  spectral domain. 

>  Current Proposed recommendation contains a version of the registry
>  section with utype on  the  ivoa schema which I think now is wrong
> "On the ivoa schema" would be wrong, right. But at least in my docs
> it's the obs_radio table that the utype is on.  And I'd argue that's
> exactly where it should be.
>  because it implies that only radio data are included in the
>  service. So I think you mean go back to your previous proposal
>  which was
> Uh... My checkout (git commit [d50ff42](https://github.com/ivoa-std/ObsCoreExtensionForRadioData/commit/d50ff42dd25c1e95ac8da48ccc0e592194ff094a)) still says

I will make a specific answer on this one

> 
>   Compliant *tables* use the utype
>   ivo://ivoa.net/std/ObsCore#radioExt-1.0.
> 
> (emphasis mine), and the sample-record.xml also has the utype at the
> right position.  If the document says the utype is on the schema
> somewhere, that's a bug and obviously needs to be fixed.
> 
> Where do you see that?  [the doc repo is down for me right now, so I
> looked at github HEAD, which I think ought to be at least as good...]

**Comment by Bonnarel** [comment]
> On Sun, Jan 12, 2025 at 02:52:23PM -0800, François Bonnarel wrote:
>  > Just to prevent possible confusion: The current draft says we

>>  Current Proposed recommendation contains a version of the registry
>>  section with utype on  the  ivoa schema which I think now is wrong
>> "On the ivoa schema" would be wrong, right. But at least in my docs
>> it's the obs_radio table that the utype is on.  And I'd argue that's
>> exactly where it should be.
>>  because it implies that only radio data are included in the
>>  service. So I think you mean go back to your previous proposal
>>  which was
> Uh... My checkout (git commit [d50ff42](https://github.com/ivoa-std/ObsCoreExtensionForRadioData/commit/d50ff42dd25c1e95ac8da48ccc0e592194ff094a)) still says
> 
>   Compliant *tables* use the utype
>   ivo://ivoa.net/std/ObsCore#radioExt-1.0.
> 
> (emphasis mine), and the sample-record.xml also has the utype at the
> right position.  If the document says the utype is on the schema
> somewhere, that's a bug and obviously needs to be fixed.
> 
> Where do you see that?  [the doc repo is down for me right now, so I
> looked at github HEAD, which I think ought to be at least as good...]

\section{Registry Aspects}
\label{sec:registry}

Services compliant with this specification are registered using
VODataService \citep{2021ivoa.spec.1102D} tablesets. Compliant tables
use the utype
$$
\hbox{\nolinkurl{ivo://ivoa.net/std/ObsCore#radioExt-1.0}.}
$$

The whole text of the registry section written between the two last interops (Sydney and Malta) states

While it is admitted that the table only sits in the tableset of the
embedding TAP service, implementors are urged to use a seperate registry
record with the main TAP service as an auxiliary capability
\citep{2019ivoa.spec.0520D}. In this way, meaningful information
on coverage in space, time, and spectrum as per VODataService 1.2 can
be communicated to the Registry, which, again, data providers are urged
to do. There is no expectation that the coverage information only
pertains to data with entries in \verb|ivoa.obscore_radio|, i.e., it may be
a copy of the coverage of the basic ObsCore table.\todo{Is that
acceptable? Or should we require pure radio coverage?}

A sample registry record for an obscore\_radio table comes with this
specification\footnote{\auxiliaryurl{sample-record.xml}}.
$$
In order to find out directly if the ivoa schema contains the radio extension
the schema MUST be defined in the registry with the utype
\hbox{\nolinkurl{ivo://ivoa.net/std/ObsCore#obscore-radioExt-1.0}}
$$
To obtain access URLs of all TAP services that have compliant tables
together with their table names (which in this major version are fixed
to \verb|ivoa.obscore_radio|), use a RegTAP \citep{2019ivoa.spec.1011D}
query like:

\begin{lstlisting}
SELECT DISTINCT(access_url), table_name
FROM rr.res_table
NATURAL JOIN rr.capability
NATURAL JOIN rr.interface
WHERE
standard_id LIKE 'ivo://ivoa.net/std/tap%'
AND intf_role='std'
AND table_utype LIKE 'ivo://ivoa.net/std/ObsCore#radioExt-%'
\end{lstlisting}

Alternatively the schema and tables can be discovered this way

\begin{lstlisting}
SELECT DISTINCT(access_url), table_name, schema_name
FROM rr.res_table
NATURAL JOIN rr.capability
NATURAL JOIN rr.interface
NATURAL JOIN rr.res_schema
WHERE 
standard_id LIKE 'ivo://ivoa.net/std/tap%'
AND intf_role='std'
AND schema_utype LIKE 'ivo://ivoa.net/std/ObsCore#obscore-radioExt-%'
\end{lstlisting}

**Comment by msdemlei** [comment]
On Mon, Jan 20, 2025 at 01:58:41AM -0800, François Bonnarel wrote:
> A sample registry record for an obscore\_radio table comes with this
> specification\footnote{\auxiliaryurl{sample-record.xml}}.
> $$
> In order to find out directly if the ivoa schema contains the radio extension
> the schema MUST be defined in the registry with the utype
> \hbox{\nolinkurl{ivo://ivoa.net/std/ObsCore#obscore-radioExt-1.0}}
> $$

Oh, yes, that's at least conceptually wrong (although it might still
work), and we shouldn't do that.  Just use the original language and
all is fine.

The Obscore utype, by the say, should also sit on the table rather
than the ivoa schema.  That's, for instance, a good way make things
future-proof in case the ivoa schema grows a second table having the
obscore columns.

**Comment by msdemlei** [comment]
On Mon, Jan 20, 2025 at 12:46:25AM -0800, François Bonnarel wrote:
> > On Sun, Jan 12, 2025 at 02:52:23PM -0800, François Bonnarel wrote:
> > I would submit that "there is a table with the #radioext-1.0 utype in
> > this TAP service" is exactly that spot.  What do you see missing from
> > it?  Or is it some duplication that you are concerned about?
>
> If you find out this table you still have to check if there is a
> root ObsCore table in the same schema

As I said: Only a prakster would publish a #radioext-1.0 table
without the corresponding ivoa.obscore.

If you want to defend against those, just say:

  You must not have a #radioext-1.0 utype on a table unless there is
  an ivoa.obscore table of version 1.1 or later in the same service.

Or something along those lines; but as I said, I consider the risk of
stand-alone radioext tables as very close to zero, so I could live
without such language.

**Comment by Zarquan** [comment]
> As I said: Only a prakster would publish a #radioext-1.0 table without the corresponding ivoa.obscore.

I can imagine a case where someone is developing a non-standard `custom-new-table`  that  can be joined with a  `#radioext-1.0` table to produce useful results. While that is in development they might want to publish/register the novel combination of `custom-new-table` and  `#radioext-1.0` tables to make it machine discoverable, without having to have a dummy, empty, and confusing, `ObsCore` table in their database.
-- proud prankster


> If you want to defend against those, just say:
>    ....

That's OK for the humans, how do we define that relationship in a machine readable way without hard coding it as a special case ? We would need to have a mechanism for expressing a _must-have_ dependency between the tables so that a generic client would automagically _just know_ that having a `#radioext-1.0` table means there is by definition an `ObsCore` table in the same schema.

Do we think of a `#radioext-1.0` table as a set of attributes **that can only ever** be used in combination with a standard `ObsCore` table ?

Or are they a set of attributes for describing radio observations that can be used to augment the content of tables such as `ObsCore` ?

Do the  `#radioext-1.0`attributes loose their meaning and become useless if they are combined with anything other than an `ObsCore` table ?

**Comment by msdemlei** [comment]
On Mon, Jan 20, 2025 at 03:00:55PM +0000, Zarquan wrote:
> `custom-new-table`  that  can be joined with a  `#radioext-1.0`
> table to produce useful results. While that is in development they
> might want to publish/register the novel combination of
> `custom-new-table` and  `#radioext-1.0` tables to make it machine
> discoverable, without having to have a dummy, empty, and confusing,

But what sort of client would want to see this odd combination of
#radioext-1.0 and custom-contrived-table *and* need to know that one
part of it is #radioext-1.0?  What action would it perform when it
has discovered it?

There's also the plain and practical point: *If* someone were devious
enough to do that, the harm would still be rather limited, viz., one
more broken service; that's routine for anything doing global
discovery.  You could say: "Well, let's reduce the pollution with
broken services".  To which I will reply: "Yup.  By avoiding
unnecessary complexity and curation of our system".

> > If you want to defend against those, just say:
> >    ....
>
> That's OK for the humans, how do we define that relationship in a
> machine readable way without hard coding it as a special case ? We

But why would a machine want to learn about this relationship in the
first place?  What would it plausibly do with such an information?

*The* only plausible use case for this identifier I see is:

  Give me all obscore services which will (if they work) be able to
  execute my query against obscore NATURAL JOIN radio_ext.

For that, just having a single utype on the table element in the
tableset works fine.  *If* we have other *discovery* use cases, let's
spell them out in a similar way, but please always do it from the
client point of view, because otherwise ("as a service writer, I
might want to express...") we'll never get anywhere.

My golden rule of VO protocols: Don't think about what you want to
say.  Think about what others want to listen to.

> Do we think of a `#radioext-1.0` table as a set of attributes
> **that can only ever** be used in combination with a standard
> `ObsCore` table ?

Yes.  Me, at least.  That's why it's called an "obscore extension".

> Or are they a set of attributes for describing radio observations
> that can be used to augment the content of tables such as `ObsCore`?

No.  Why should they?  *If* you build custom tables with radio
"attributes", you are probably well-advised to re-use these columns,
just as all kinds of other standards re-use columns derived from
obscore.  But that's for the convenience of humans who don't
appreciate having to learn different names for the same thing (and
yet always invent those).

In other words: I don't think there is a credible thing that a
machine could use an information like "this column metadata snarfed
from obscore" for.

> Do the  `#radioext-1.0`attributes lose their meaning and become
> useless if they are combined with anything other than an `ObsCore`
> table?

Well, they certainly can't be sensibly used any more.  If you don't
know where you looked, there's precious little you can do with the
information that your frequency bins are 10 kHz...

**Comment by Bonnarel** [comment]
Hi, how to discover the various ObsCore tables for the root set of Obscore columns and for different specific extensions 

If the user wants to know if the service contains the root ObsCore table she could query regtap this way : 

\begin{lstlisting}
SELECT DISTINCT(access_url), table_name
FROM rr.res_table
NATURAL JOIN rr.capability
NATURAL JOIN rr.interface
WHERE
standard_id LIKE 'ivo://ivoa.net/std/tap%'
AND intf_role='std'
AND table_utype LIKE'ivo://ivoa.net/std/ObsCore#core-1.1'
\end{lstlisting} 
----> actually it doesn't work this way today  but it is the way it should work in the future

**Comment by msdemlei** [comment]
On Mon, Jan 20, 2025 at 02:43:05PM -0800, François Bonnarel wrote:
> WHERE
> standard_id LIKE 'ivo://ivoa.net/std/tap%'
> AND intf_role='std'
> AND table_utype LIKE'ivo://ivoa.net/std/ObsCore#core-1.1'

Quick correction: We want to discover "all things of version 1" to
allow minor version changes without breaking things, so the last
constraint ought to be

AND table_utype LIKE 'ivo://ivoa.net/std/ObsCore#core-1.%'

But thanks for pointing this out, and if anyone reading this
publishes Obscore tables: Please give them their utype right now
already.  It'd be break if we could migrate to the new discovery
pattern even before Obscore 1.2.

**Comment by msdemlei** [comment]
On Thu, Feb 13, 2025 at 04:47:51PM +0000, François Bonnarel wrote:
> >avoid but are part of the whole, the service has just chosen to
> >not implement those parts.  Perhaps this ‘optionality’ is what
> >people mean by ‘extension’?

It would seem to me that "Optional" (DB lingo: may be NULL) and
"Domain-specific" (DB lingo: a set of columns you can use in your
query) are distinct, although not necessarily orthogonal, concepts.
I believe it is the slight entanglement between these two concpets
that makes this whole thing a bit troublesome.

Still, I think in order to figure out how to design the obscore
extension mechanism, it pays to strictly take a user position.  And
for them, I would argue we have to satisfactorily answer two
questions:

(a) how to I find services I can send my queries to?

(b) can I send identical queries to all of them and have a fair
chance that the queries will have roughly identical semantics on all
of them.

I would argue that separate extension tables with fixed column
sets intended to be NATURAL-ly JOIN-ed together (i.e., my original
proposal) are the best bet we have to have good and workable answers
to both.

**Comment by Bonnarel** [comment]
> On Thu, Feb 13, 2025 at 04:47:51PM +0000, François Bonnarel wrote:
>  >avoid but are part of the whole, the service has just chosen to
>  >not implement those parts.  Perhaps this ‘optionality’ is what
>  >people mean by ‘extension’?
> It would seem to me that "Optional" (DB lingo: may be NULL) and
> "Domain-specific" (DB lingo: a set of columns you can use in your
> query) are distinct, although not necessarily orthogonal, concepts.
> I believe it is the slight entanglement between these two concpets
> that makes this whole thing a bit troublesome.
> 
> Still, I think in order to figure out how to design the obscore
> extension mechanism, it pays to strictly take a user position.  And
> for them, I would argue we have to satisfactorily answer two
> questions:
> 
> (a) how to I find services I can send my queries to?
> 
> (b) can I send identical queries to all of them and have a fair
> chance that the queries will have roughly identical semantics on all
> of them.
> 
> I would argue that separate extension tables with fixed column
> sets intended to be NATURAL-ly JOIN-ed together (i.e., my original
> proposal) are the best bet we have to have good and workable answers
> to both.

Following what I wrote 2 weeks ago, and after discussion among authors, I removed anything related to standardID.
So we have two tables ObsCore and obscore_radio in the same ivoa schema.
Discovering services delivering the extension can be done on the table name at the moment. 
standardID for the various extensions will be defined within ObscOre 1.2 (or 2.0)

**Comment by msdemlei** [comment]
On Thu, Feb 27, 2025 at 05:37:27AM -0800, François Bonnarel wrote:
> Following what I wrote 2 weeks ago, and after discussion among
> authors, I removed anything related to standardID.

Uh, ouch.

> So we have two tables ObsCore and obscore_radio in the same ivoa schema.
> Discovering services delivering the extension can be done on the
> table name at the moment.

Ouch again: While I give you that we perhaps *should* have designed
things in this way, we have not.  It would suck if obs_radio would
now try something new and different from what everything else does
(except if there was a strong reason for deviating from existing
practices, which I think there is not).

> standardID for the various extensions will be defined within
> ObscOre 1.2 (or 2.0)

Aw, let's not do that.  That would mean we have to change Obscore
every time there's a new extension.

What's your motivation for these changes?

See also <http://ivoa.net/documents/Notes/TableReg> for how table
discovery should be unified in the future, taking into account the
experiences with obscore, epntap, and their friends.

**Comment by Bonnarel** [comment]
> On Thu, Feb 27, 2025 at 05:37:27AM -0800, François Bonnarel wrote:
>  Following what I wrote 2 weeks ago, and after discussion among
>  authors, I removed anything related to standardID.
> Uh, ouch.

Sorry for the delay
See below

>  So we have two tables ObsCore and obscore_radio in the same ivoa schema.
>  Discovering services delivering the extension can be done on the
>  table name at the moment.
> Ouch again: While I give you that we perhaps *should* have designed
> things in this way, we have not.  It would suck if obs_radio would
> now try something new and different from what everything else does
> (except if there was a strong reason for deviating from existing
> practices, which I think there is not).

Don't be afraid. I am not proposing some new general way of discovering ObsCore tables. My proposal is only provisional because apparently the ObsCore extension for radio wil be recommanded by a simple endorsed note, while IVOA prepares a global rework of ObsCore which will encompass the various extensions 

>  standardID for the various extensions will be defined within
>  ObscOre 1.2 (or 2.0)
> Aw, let's not do that.  That would mean we have to change Obscore
> every time there's a new extension

The extension for radio is a kind of precursor for several extensions. The new root ObscOre and extensions could be defined in the future version of ObsCore. That would be the right place to define standardId for the various extensions and where to put them. 
.
> 
> What's your motivation for these changes?
> 

See above

> See also <http://ivoa.net/documents/Notes/TableReg> for how table
> discovery should be unified in the future, taking into account the
> experiences with obscore, epntap, and their friends.

As you read me before in this issue you know that I'm aware of this note. But on which table utype should we set the standarID, that's wher we differ and this is strongly related to the status of the extensions with respect to ObsCore. So
that's why it's better to wait I think.
